subsys/bluetooth/audio/csip_set_member.c

```diff
@@ -722,8 +771,28 @@ int bt_csip_set_member_register(const struct bt_csip_set_member_register_param *
 	inst->service_p = &csip_set_member_service_list[instance_cnt];
 	instance_cnt++;
 
-	bt_conn_cb_register(&conn_callbacks);
-	bt_conn_auth_info_cb_register(&auth_callbacks);
+	if (!callbacks_registered) {
+		bt_conn_cb_register(&conn_callbacks);
+		bt_conn_auth_info_cb_register(&auth_callbacks);
+
+		callbacks_registered = true;
+	}
+
+	/* The removal of the optional characteristics should be done in reverse order of the order
+	 * in BT_CSIP_SERVICE_DEFINITION, as that improves the performance of remove_csis_char,
+	 * since it's easier to remove the last characteristic
+	 */
+	if (param->rank == 0U) {
+		remove_csis_char(BT_UUID_CSIS_RANK, inst->service_p);
+	}
+
+	if (param->set_size == 0U) {
+		remove_csis_char(BT_UUID_CSIS_SET_SIZE, inst->service_p);
+	}
+
+	if (!param->lockable) {
+		remove_csis_char(BT_UUID_CSIS_SET_LOCK, inst->service_p);
+	}
```

Comment:
Wouldn't it be easier and cleaner to do the opposite? I mean to compose the service attributes based on parameters given?
Luckily, there is no `bt_csip_set_member_unregister` method, because, currently, this does not restore the `csip_set_member_service_list` initial state.

Reply:
That's a good point, but the reason why it is done this way, is that the list of CSIS instances are declared using `BT_GATT_SERVICE_INSTANCE_DEFINE`. 

We could modify `BT_CSIP_SERVICE_DEFINITION` to have these 3 characteristics as empty by default and then populate them when requested. That might indeed by the cleaner solution, even if takes a bit more effort :) 

Let me know if that makes sense to you